airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ProtoToCsvWriter.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2025 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.file.object_storage
+
+import com.fasterxml.jackson.core.io.JsonStringEncoder
+import io.airbyte.cdk.load.data.*
+import io.airbyte.cdk.load.message.*
+import java.time.Instant
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+
+class ProtoToCsvWriter(
+    private val header: Array<String>,
+    private val columns: Array<AirbyteValueProxy.FieldAccessor>,
+    private val rootLevelFlattening: Boolean,
+    private val extractedAtTsTz: Boolean,
+) {
+
+    private val idxRawId = header.indexOf(Meta.COLUMN_NAME_AB_RAW_ID)
+    private val idxExtractedAt = header.indexOf(Meta.COLUMN_NAME_AB_EXTRACTED_AT)
+    private val idxMeta = header.indexOf(Meta.COLUMN_NAME_AB_META)
+    private val idxGenerationId = header.indexOf(Meta.COLUMN_NAME_AB_GENERATION_ID)
+    private val idxData = header.indexOf(Meta.COLUMN_NAME_DATA)
+    private val jsonEscaper = JsonStringEncoder.getInstance()
+
+    private val rowIndex: Map<String, Int> =
+        if (rootLevelFlattening) {
+            columns.associate { it.name to header.indexOf(it.name) }
+        } else {
+            emptyMap()
+        }
+
+    private val rowBuf: Array<Any?> = arrayOfNulls(header.size)
```

Comment:
why do we need to preallocate this array? Are there any potential risks with race conditions (e.g. multiple threads calling `toCsvRow` simultaneously)